src/main/java/org/zalando/planb/provider/AuthorizeController.java

```diff
@@ -76,97 +130,273 @@ String showAuthorizationForm(@RequestParam(value = "realm") Optional<String> rea
+                redirect = generateRedirectURIbasedOnGrantType(responseType, realmName, userRealm, state, clientId, finalScopes, claims, redirectUri);
             }
+
         } catch (UserRealmAuthenticationException e) {
-            log.info("{} (status {} / {})", e.getMessage(), e.getStatusCode(), e.getClass().getSimpleName());
             // redirect back to login form with error message
-            redirect = new URIBuilder("/oauth2/authorize")
-                    .addParameter("response_type", responseType)
-                    .addParameter("realm", realmName)
-                    .addParameter("client_id", clientId)
-                    .addParameter("scope", ScopeProperties.join(finalScopes))
-                    .addParameter("redirect_uri", redirectUri.toString())
-                    .addParameter("state", state.orElse(""))
-                    .addParameter("error", "access_denied")
-                    .build();
+            log.info("{} (status {} / {})", e.getMessage(), e.getStatusCode(), e.getClass().getSimpleName());
+            redirect = generateLoginFormURI(realmName, responseType, state, clientId, finalScopes, redirectUri);
+        }
+
+        authorizeResponse.setRedirect(redirect.toString());
+        return authorizeResponse;
+    }
+
+    @RequestMapping(method = RequestMethod.POST)
+    ModelAndView authorizeAsModel(
+            @RequestParam(value = RESPONSE_TYPE_SNAKECASE, required = true) final String responseType,
+            @RequestParam(value = REALM) final Optional<String> realmNameParam,
+            @RequestParam(value = CLIENT_ID_SNAKECASE, required = true) final String clientId,
+            @RequestParam(value = SCOPE) final Optional<String> scope,
+            @RequestParam(value = REDIRECT_URI_SNAKECASE) final URI redirectUri,
+            @RequestParam(value = STATE) final Optional<String> state,
+            @RequestParam(value = USERNAME, required = true) final String username,
+            @RequestParam(value = PASSWORD, required = true) final String password,
+            @RequestParam(value = DECISION) final Optional<String> decision,
+            @RequestHeader(value = HOST) final Optional<String> hostHeader) throws IOException, URISyntaxException,
+        JOSEException {
+        final AuthorizeResponse authorizeResponse = authorizeAsJson(responseType, realmNameParam, clientId, scope,
+                redirectUri, state, username, password, decision, hostHeader);
+        ModelAndView modelAndView;
+
+        if (authorizeResponse.isConsentNeeded()) {
+            modelAndView = generateConsentScopesView(authorizeResponse);
+        } else {
+            modelAndView = new ModelAndView(new RedirectView(authorizeResponse.getRedirect()));
+        }
+
+        return modelAndView;
+    }
+
+    private boolean allScopesAreConsented(final Optional<Set<String>> consentedScopes, final Set<String> finalScopes) {
+        return consentedScopes.get().containsAll(finalScopes);
```

Comment:
it is dangerous to invoke `.get()` of an Optional, without checking if it `isPresent()` before. This will produce a RuntimeException in case it's empty. Better always use `map` and the `orElse` methods when dealing with Optionals. E.g. `consentedScopes.map(set -> set.containsAll(finalScopes)).orElse(true)`. But in this case I would just change the signature of this private helper method to accept a Set instead of an Optional.